pkgs/build-support/mkshellwrapper/default.nix

```diff
@@ -0,0 +1,95 @@
+{ mkShell
+, writeScript
+, bashInteractive
+, lib
+, writeText
+, coreutils
+}:
+
+# This function creates a wrapper script that runs a command in a shell context.
+{ # The shell that will execute the wrapper script
+  shell ? bashInteractive + "/bin/bash"
+, # The result of calling mkShell
+  drv
+, # Which commands to run before the payload call
+  prelude ? ""
+, ...
+}:
+
+let
+  shellDrv = if lib.isDerivation drv then drv else mkShell drv;
+
+  # This function closely mirrors what this Nix code does:
+  # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/primops.cc#L1102
+  # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/eval.cc#L1981-L2036
+  stringValue = value:
+    # We can't just use `toString` on all derivation attributes because that
+    # would not put path literals in the closure. So we explicitly copy
+    # those into the store here
+    if builtins.typeOf value == "path" then "${value}"
+    else if builtins.typeOf value == "list" then toString (map stringValue value)
+    else toString value;
+
+  # A binary that calls the command to build the derivation
+  builder = writeScript "buildDerivation" ''
+    exec ${lib.escapeShellArg (stringValue shellDrv.drvAttrs.builder)} ${lib.escapeShellArgs (map stringValue shellDrv.drvAttrs.args)}
+  '';
+
+  # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L992-L1004
+  drvEnv = lib.mapAttrs' (name: value:
+    let str = stringValue value;
+    in if lib.elem name (shellDrv.drvAttrs.passAsFile or [])
+    then lib.nameValuePair "${name}Path" (writeText "pass-as-text-${name}" str)
+    else lib.nameValuePair name str
+  ) shellDrv.drvAttrs //
+    # A mapping from output name to the nix store path where they should end up
+    # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/primops.cc#L1253
+    lib.genAttrs shellDrv.outputs (output: builtins.unsafeDiscardStringContext shellDrv.${output}.outPath);
+
+  staticPath = "${dirOf shell}:${lib.makeBinPath [ builder coreutils ]}";
+
+  env = drvEnv // {
+    # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1027-L1030
+    # PATH = "/path-not-set";
+    # Allows calling bash and `buildDerivation` as the Cmd
+    PATH = staticPath;
+
```

Comment:
Drop this empty line as this is not required.

Reply:
Done!